In [6]:
import pandas as pd
import numpy as np

tsv_file='..//data//NSDUH_2017.tsv'
csv_table=pd.read_table(tsv_file,sep='\t', low_memory=False)

In [7]:
#csv_table.to_csv('nsduh_backup.csv',index=False)
csv_table.head()
main_data = csv_table

In [8]:
categories_file = '..//data//categories.csv'
categories = pd.read_csv(categories_file, sep=';')
categories.head()

Category    Column  \
0  INTERVIEW INFORMATION   QUESTID2   
1  INTERVIEW INFORMATION   FILEDATE   
2          SUBSTANCE USE    CIGEVER   
3          SUBSTANCE USE   CIGOFRSM   
4          SUBSTANCE USE   CIGWILYR   

                                    Description    Flag  
0                     RESPONDENT IDENTIFICATION  DELETE  
1                     CREATION DATE OF THE DATA  DELETE  
2                       EVER SMOKED A CIGARETTE    GOOD  
3   IF BEST FRIEND OFFERED, WOULD YOU SMOKE CIG    GOOD  
4  DO YOU THINK YOU'LL SMOKE CIG IN NEXT 12 MOS    GOOD

In [9]:
main_data.head()

QUESTID2    FILEDATE  CIGEVER  CIGOFRSM  CIGWILYR  CIGTRY  CIGYFU  CIGMFU  \
0  55235143  10/09/2018        1        99        99      13    9999      99   
1  13435143  10/09/2018        1        99        99      15    9999      99   
2  81345143  10/09/2018        1        99        99      14    9999      99   
3  53955143  10/09/2018        1        99        99      16    9999      99   
4  51775143  10/09/2018        2        99        99     991    9991      91   

   CIGREC  CIG30USE  ...    POVERTY3  TOOLONG  TROUBUND  PDEN10  COUTYP4  \
0       4        93  ...         3.0        2         2       1        1   
1       1        18  ...         3.0        1         2       1        1   
2       1        10  ...         3.0        2         2       1        1   
3       4        93  ...         3.0        2         2       2        2   
4      91        91  ...         3.0        1         1       1        1   

   MAIIN102  AIIND102      ANALWT_C  VESTR  VEREP  
0         2         2  11203.888954  40043      1  
1         2         2   9496.462244  40006      2  
2         2         2   2943.702802  40030      2  
3         2         2   1783.702549  40026      2  
4         2         2  31528.749357  40029      1  

[5 rows x 2668 columns]

In [10]:
delete_categories = categories.loc[categories.Flag == 'DELETE']
delete_categories
delete_categories_column = delete_categories['Column']
delete_categories_list = delete_categories_column.values.tolist()
delete_categories_list

['QUESTID2',
 'FILEDATE',
 'IICIGRC',
 'II2CIGRC',
 'IICGRRC',
 'II2CGRRC',
 'IIPIPLF',
 'IIPIPMN',
 'IISMKLSSREC',
 'IIALCRC',
 'II2ALCRC',
 'IIMJRC',
 'II2MJRC',
 'IICOCRC',
 'II2COCRC',
 'IICRKRC',
 'II2CRKRC',
 'IIHERRC',
 'II2HERRC',
 'IIHALLUCREC',
 'IILSDRC',
 'II2LSDRC',
 'IIPCPRC',
 'II2PCPRC',
 'IIECSTMOREC',
 'IIKETMINREC',
 'IIDAMTFXREC',
 'IISALVIAREC',
 'IIINHALREC',
 'IIMETHAMREC',
 'IIPNRANYREC',
 'IIOXCNANYYR',
 'IITRQANYREC',
 'IISTMANYREC',
 'IISEDANYREC',
 'IIPNRNMREC',
 'IIOXCNNMYR',
 'IITRQNMREC',
 'IISTMNMREC',
 'IISEDNMREC',
 'IIALCFY',
 'II2ALCFY',
 'IIMJFY',
 'II2MJFY',
 'IICOCFY',
 'II2COCFY',
 'IICRKFY',
 'II2CRKFY',
 'IIHERFY',
 'II2HERFY',
 'IIHALLUCYFQ',
 'IIINHALYFQ',
 'IIMETHAMYFQ',
 'IICIGFM',
 'II2CIGFM',
 'IICGRFM',
 'II2CGRFM',
 'IISMKLSS30N',
 'IIALCFM',
 'II2ALCFM',
 'IIALCBNG30D',
 'IIMJFM',
 'II2MJFM',
 'IICOCFM',
 'II2COCFM',
 'IICRKFM',
 'II2CRKFM',
 'IIHERFM',
 'II2HERFM',
 'IIHALLUC30N',
 'IIINHAL30N',
 'IIMETHAM30N',
 'IIPNRNM30FQ',
 'IITRQ

In [11]:
good_main_data = main_data.drop(columns=delete_categories_list)
good_main_data.head()

CIGEVER  CIGOFRSM  CIGWILYR  CIGTRY  CIGYFU  CIGMFU  CIGREC  CIG30USE  \
0        1        99        99      13    9999      99       4        93   
1        1        99        99      15    9999      99       1        18   
2        1        99        99      14    9999      99       1        10   
3        1        99        99      16    9999      99       4        93   
4        2        99        99     991    9991      91      91        91   

   CG30EST  CIG30AV    ...     IRFAMIN3  GOVTPROG  INCOME  POVERTY3  TOOLONG  \
0       93       93    ...            7         2       4       3.0        2   
1       99        4    ...            4         2       2       3.0        1   
2       99        3    ...            7         2       4       3.0        2   
3       93       93    ...            7         2       4       3.0        2   
4       91       91    ...            5         2       2       3.0        1   

   TROUBUND  PDEN10  COUTYP4  MAIIN102  AIIND102  
0         2       1        1         2         2  
1         2       1        1         2         2  
2         2       1        1         2         2  
3         2       2        2         2         2  
4         1       1        1         2         2  

[5 rows x 2492 columns]

In [12]:
good_main_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56276 entries, 0 to 56275
Columns: 2492 entries, CIGEVER to AIIND102
dtypes: float64(406), int64(2086)
memory usage: 1.0 GB


In [ ]:
#good_main_data.to_csv('..\\data\\nsduh_clean.csv',index=False)

In [13]:
nulls_in_columns = good_main_data.isnull().any()
nulls_in_columns

CIGEVER      False
CIGOFRSM     False
CIGWILYR     False
CIGTRY       False
CIGYFU       False
CIGMFU       False
CIGREC       False
CIG30USE     False
CG30EST      False
CIG30AV      False
CIG30BR2     False
CIG30TPE     False
CIG30MEN     False
CIG30MLN     False
CIG30RO2     False
CIGDLYMO     False
CIGAGE       False
CIGDLYFU     False
CIGDLMFU     False
CIG100LF     False
SMKLSSEVR    False
SMKLSSTRY    False
SMKLSSYFU    False
SMKLSSMFU    False
SMKLSSREC    False
SMKLSS30N    False
SMKLSS30E    False
CIGAREVR     False
CIGARTRY     False
CIGARYFU     False
             ...  
HLNVSOR      False
IRMCDCHP     False
IRMEDICR     False
IRCHMPUS     False
IRPRVHLT     False
IROTHHLT     False
HLCALLFG     False
HLCALL99     False
ANYHLTI2     False
IRINSUR4     False
OTHINS       False
CELLWRKNG    False
CELLNOTCL    False
IRFAMSOC     False
IRFAMSSI     False
IRFSTAMP     False
IRFAMPMT     False
IRFAMSVC     False
IRWELMOS     False
IRPINC3      False
IRFAMIN3     False
GOVTPROG    

In [14]:
rows_nan = good_main_data.isnull().sum(axis = 1)
rows_nan_count = len([i for i in rows_nan if i != 0])
rows_nan_count

56276

Most of rows have NaNs

In [15]:
columns_nan = good_main_data.isnull().sum(axis = 0)
columns_nan_count = len([(x, i) for x, i in enumerate(columns_nan) if i != 0])
columns_nan_count

376

The dataset has 376 columns which contains NaN values. Delete all columns with more than 10% NaNs which is the ones with more than 5628 NaNs. It'll be 310 columns. The rest 66 of them will be replaced with most common value in that column.

In [16]:
columns_nan_count = len([(x, i) for x, i in enumerate(columns_nan) if i >= 5628])
columns_nan_count

310

In [17]:
columns_nan_list = [x for x, i in enumerate(columns_nan) if i >= 5628]
columns_nan_list

[636,
 637,
 638,
 639,
 682,
 683,
 684,
 685,
 686,
 687,
 688,
 689,
 690,
 691,
 692,
 693,
 1085,
 1461,
 1462,
 1463,
 1464,
 1465,
 1466,
 1467,
 1468,
 1469,
 1470,
 1471,
 1472,
 1473,
 1474,
 1475,
 1626,
 1627,
 1628,
 1629,
 1630,
 1631,
 1632,
 1633,
 1634,
 1635,
 1636,
 1637,
 1638,
 1639,
 1640,
 1641,
 1642,
 1643,
 1644,
 1645,
 1646,
 1647,
 1648,
 1649,
 1650,
 1651,
 1652,
 1653,
 1654,
 1655,
 1656,
 1657,
 1658,
 1659,
 1660,
 1661,
 1662,
 1663,
 1664,
 1665,
 1666,
 1667,
 1668,
 1669,
 1670,
 1740,
 1741,
 1742,
 1743,
 1744,
 1745,
 1746,
 1747,
 1748,
 1749,
 1750,
 1751,
 1752,
 1753,
 1754,
 1755,
 1756,
 1757,
 1758,
 1759,
 1760,
 1761,
 1762,
 1763,
 1764,
 1765,
 1766,
 1767,
 1768,
 1769,
 1770,
 1771,
 1772,
 1773,
 1774,
 1775,
 1776,
 1777,
 1778,
 1779,
 1780,
 1781,
 1782,
 1783,
 1784,
 1785,
 1786,
 1819,
 1820,
 1821,
 1822,
 1823,
 1824,
 1825,
 1826,
 1827,
 1828,
 1829,
 1830,
 1831,
 1832,
 1833,
 1834,
 1835,
 1836,
 1837,
 1838,
 1839,
 

In [18]:
without_nan_main_data = good_main_data.drop(good_main_data.columns[columns_nan_list],axis=1)

In [19]:
without_nan_main_data.head()

CIGEVER  CIGOFRSM  CIGWILYR  CIGTRY  CIGYFU  CIGMFU  CIGREC  CIG30USE  \
0        1        99        99      13    9999      99       4        93   
1        1        99        99      15    9999      99       1        18   
2        1        99        99      14    9999      99       1        10   
3        1        99        99      16    9999      99       4        93   
4        2        99        99     991    9991      91      91        91   

   CG30EST  CIG30AV    ...     IRFAMIN3  GOVTPROG  INCOME  POVERTY3  TOOLONG  \
0       93       93    ...            7         2       4       3.0        2   
1       99        4    ...            4         2       2       3.0        1   
2       99        3    ...            7         2       4       3.0        2   
3       93       93    ...            7         2       4       3.0        2   
4       91       91    ...            5         2       2       3.0        1   

   TROUBUND  PDEN10  COUTYP4  MAIIN102  AIIND102  
0         2       1        1         2         2  
1         2       1        1         2         2  
2         2       1        1         2         2  
3         2       2        2         2         2  
4         1       1        1         2         2  

[5 rows x 2182 columns]

In [20]:
columns_nan_check = without_nan_main_data.isnull().sum(axis = 0)
columns_nan_count_check = len([(x, i) for x, i in enumerate(columns_nan_check) if i != 0])
columns_nan_count_check

66

In [21]:
replace_nan_data = without_nan_main_data.fillna(without_nan_main_data.mode().iloc[0])
replace_nan_data.head()

CIGEVER  CIGOFRSM  CIGWILYR  CIGTRY  CIGYFU  CIGMFU  CIGREC  CIG30USE  \
0        1        99        99      13    9999      99       4        93   
1        1        99        99      15    9999      99       1        18   
2        1        99        99      14    9999      99       1        10   
3        1        99        99      16    9999      99       4        93   
4        2        99        99     991    9991      91      91        91   

   CG30EST  CIG30AV    ...     IRFAMIN3  GOVTPROG  INCOME  POVERTY3  TOOLONG  \
0       93       93    ...            7         2       4       3.0        2   
1       99        4    ...            4         2       2       3.0        1   
2       99        3    ...            7         2       4       3.0        2   
3       93       93    ...            7         2       4       3.0        2   
4       91       91    ...            5         2       2       3.0        1   

   TROUBUND  PDEN10  COUTYP4  MAIIN102  AIIND102  
0         2       1        1         2         2  
1         2       1        1         2         2  
2         2       1        1         2         2  
3         2       2        2         2         2  
4         1       1        1         2         2  

[5 rows x 2182 columns]

In [22]:
replace_nan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56276 entries, 0 to 56275
Columns: 2182 entries, CIGEVER to AIIND102
dtypes: float64(96), int64(2086)
memory usage: 936.8 MB


In [23]:
columns_nan_check1 = replace_nan_data.isnull().sum(axis = 0)
columns_nan_count_check1 = len([(x, i) for x, i in enumerate(columns_nan_check1) if i == 0])
columns_nan_count_check1

2182

In [ ]:
#replace_nan_data.to_csv('..\\data\\nsduh_clean.csv',index=False)

In [49]:
# replace_nan_data['CRDAYPMO']